File: src/main/java/at/searles/parsing/annotation/AnnotationParser.java
package at.searles.parsing.annotation;

import at.searles.parsing.Parser;
import at.searles.parsing.ParserStream;
import at.searles.parsing.printing.ConcreteSyntaxTree;

public class AnnotationParser<C, T> implements Parser<T> {

    private final C annotation;
    private final Parser<T> parser;

    public AnnotationParser(C annotation, Parser<T> parser) {
        this.annotation = annotation;
        this.parser = parser;
    }

    @Override
    public T parse(ParserStream stream) {
        stream.notifyAnnotationBegin(annotation);
        T value = parser.parse(stream);
        stream.notifyAnnotationEnd(annotation, value != null);
        return value;
    }

    @Override
    public boolean recognize(ParserStream stream) {
        stream.notifyAnnotationBegin(annotation);
        boolean success = parser.recognize(stream);
        stream.notifyAnnotationEnd(annotation, success);
        return success;
    }

    @Override
    public ConcreteSyntaxTree print(T t) {
        ConcreteSyntaxTree output = parser.print(t);

        if (output == null) {
            return null;
        }

        return output.annotate(annotation);
    }

    @Override
    public String toString() {
        return parser.toString();
    }
}

File: src/main/java/at/searles/parsing/annotation/AnnotationParser.kt
package at.searles.parsing.annotation

import at.searles.parsing.Parser
import at.searles.parsing.ParserStream
import at.searles.parsing.printing.ConcreteSyntaxTree

class AnnotationParser<C, T>(private val annotation: C, private val parser: Parser<T>) : Parser<T> {
    override fun parse(stream: ParserStream): T? {
        stream.notifyAnnotationBegin(annotation)
        return parser.parse(stream).also {
            if(it != null) {
                stream.notifyAnnotationEndSuccess(annotation)
            } else {
                stream.notifyAnnotationEndFail(annotation)
            }
        }
    }

    override fun recognize(stream: ParserStream): Boolean {
        stream.notifyAnnotationBegin(annotation)
        return if(parser.recognize(stream)) {
            stream.notifyAnnotationEndSuccess(annotation)
            true
        } else {
            stream.notifyAnnotationEndFail(annotation)
            false
        }
    }

    override fun print(item: T): ConcreteSyntaxTree? {
        return parser.print(item)?.annotate(annotation)
    }

    override fun toString(): String {
        return parser.toString()
    }

}